test/test_algos.py

```diff
@@ -130,16 +103,54 @@ def test_apgd():
                             rel_error=None,
                         )
                         recon.set_data(data)
-                        res = recon.apply(n_iter=n_iter, disp_iter=None, plot=False)
+                        res = recon.apply(n_iter=_n_iter, disp_iter=None, plot=False)
                         assert psf.shape[3] == (1 if gray else 3)
                         assert res.dtype == psf.dtype, f"Got {res.dtype}, expected {dtype}"
     else:
         print("Pycsou not installed. Skipping APGD test.")
 
 
-if __name__ == "__main__":
-    test_gradient_descent()
-    test_admm()
-    test_nesterov_gradient_descent()
-    test_fista()
-    test_apgd()
+#  trainable algorithms
+@pytest.mark.parametrize("algorithm", trainable_algos)
+def test_trainable_recon(algorithm):
+    if torch_is_available:
+        for dtype, torch_type in [("float32", torch.float32), ("float64", torch.float64)]:
+            psf = torch.rand(1, 32, 64, 3, dtype=torch_type)
+            data = torch.rand(2, 1, 32, 64, 3, dtype=torch_type)
+            recon = UnrolledFISTA(psf, n_iter=_n_iter, dtype=dtype)
+
+            assert (
+                next(recon.parameters(), None) is not None
+            ), f"{algorithm.__name__} has no trainable parameters"
+
+            res = recon.batch_call(data)
+            loss = torch.mean(res)
+            loss.backward()
+
+            assert (
+                data.shape[0] == res.shape[0]
+            ), f"Batch dimension changed: got {res.shape[0]} expected {data.shape[0]}"
+
+            assert len(psf.shape) == 4
+            print(res.shape)
+            assert res.shape[4] == 3, "Input in HWC format but output CHW format"
+
+
+@pytest.mark.parametrize("algorithm", trainable_algos)
+def test_trainable_ind(algorithm):
+    if not torch_is_available:
+        return
+    for dtype, torch_type in [("float32", torch.float32), ("float64", torch.float64)]:
+        psf = torch.rand(1, 34, 64, 3, dtype=torch_type)
+        data1 = torch.rand(5, 1, 34, 64, 3, dtype=torch_type)
+        data2 = torch.rand(1, 1, 34, 64, 3, dtype=torch_type)
+        data2[0, 0, ...] = data1[0, 0, ...]
+
+        recon = algorithm(psf, dtype=dtype, n_iter=_n_iter)
+        res1 = recon.batch_call(data1)
+        res2 = recon.batch_call(data2)
+
+        torch.testing.assert_close(res1[0, 0, ...], res2[0, 0, ...])
```

Comment:
Would they be equivalent if data2 didn't have a batch dimension? That might be a better as separate test and keep this one fir checking that number if iterations resets

Reply:
Not sure I understand. Even if data2 has no batch dim, the result will have one (unless we call apply)